File: lab2/lab2.py
"""
Необходимо написать скрипт, обрабатывающий лог файл Nginx и выводящий список IP адресов, с которых производились запросы. Адреса из общей подсети \24 необходимо группировать при выводе (например, 10.40.0.4 и 10.40.0.231 относятся к одной подсети).

Версия Python: 3.5
"""

import ipaddress

MASK = 24

f = open("access.log", "r")
total = 0
unique = 0
networks = dict()
ips = dict()
for line in f.readlines():
	# IP address string
	ip = line.split(" - - ")[0]
	# Containing network
	net = ipaddress.ip_network(ip + '/' + str(MASK), strict=False)
	# Remember this network and this particular IP address
	networks.setdefault(net, set())
	ips.setdefault(ip, 0)
	# Adjust counters
	unique += ip not in networks[net]
	total += 1
	networks[net].add(ip)
	ips[ip] += 1
	

print("Total: ", total)
print("Unique: ", unique)
print()

for k in networks:
	if len(networks[k]) == 1:
		ip = networks[k].pop()
		print(ip, " (" + str(ips[ip]) + ")")
	else:
		print(k)
		for ip in networks[k]:
			print("\t==> " + ip +  " (" + str(ips[ip]) + ")")

File: lab2/lab2.rb
#!/usr/bin/ruby
require 'set'

networks = Hash.new
total = 0
f = File.open('access.log')

f.read.each_line do |line|
	ip = line.split(" - - ")[0]
	ip_parts = ip.split(".")
	net = ip_parts[0] + "." + ip_parts[1] + "." + ip_parts[2] + ".0"
	if networks.include?(net) == false then
		networks[net] = Set.new()
	end
	networks[net].add(ip)
	total += 1
end
print("Total: #{total}\n")
print("Unique: #{networks.values.map{ |x| x.count }.reduce(:+)}\n")
print("\n")

networks.each do |net, ips|
	print("#{net}:\n\t#{ips.to_a}\n")
end

f.close()
